op-batcher/batcher/channel_builder.go

```diff
@@ -398,46 +404,51 @@ func (c *ChannelBuilder) Close() {
 }
 
 // TotalFrames returns the total number of frames that were created in this channel so far.
-// It does not decrease when the frames queue is being emptied.
 func (c *ChannelBuilder) TotalFrames() int {
 	return c.numFrames
 }
 
 // HasFrame returns whether there's any available frame. If true, it can be
-// popped using NextFrame().
+// dequeued using NextFrame().
 //
 // Call OutputFrames before to create new frames from the channel out
 // compression pipeline.
 func (c *ChannelBuilder) HasFrame() bool {
-	return len(c.frames) > 0
+	return c.frames.Len() > 0 && c.frameCursor < c.frames.Len()
 }
 
 // PendingFrames returns the number of pending frames in the frames queue.
 // It is larger zero iff HasFrame() returns true.
 func (c *ChannelBuilder) PendingFrames() int {
-	return len(c.frames)
+	return c.frames.Len() - c.frameCursor
 }
 
-// NextFrame dequeues the next available frame.
+// NextFrame returns the next available frame and increments the frameCursor
 // HasFrame must be called prior to check if there's a next frame available.
 // Panics if called when there's no next frame.
 func (c *ChannelBuilder) NextFrame() frameData {
-	if len(c.frames) == 0 {
+	if len(c.frames) <= c.frameCursor {
 		panic("no next frame")
```

Comment:
Probably an enhancement (maybe we should write up an issue) -- this (and elsewhere there are panics) feels more like an error that should be handled gracefully and logged/metric pushed as opposed to crashing the program? Or is there a reason why we want to `panic`?

Reply:
Interesting one. I agree that this could probably be reworked to remove the need for a panic. It's not a great devex to say "you must do this other thing before calling this method or it will panic". The two methods could be combined most likely.

Reply:
Filed an issue, could be a fast follow: https://github.com/ethereum-optimism/optimism/issues/12742